crawl-ref/source/stairs.cc

```diff
@@ -464,11 +480,18 @@ static level_id _travel_destination(const dungeon_feature_type how,
     // Falling down is checked before the transition if going upstairs, since
     // it might prevent the transition itself.
     if (going_up && _check_fall_down_stairs(how, true))
+    {
         // TODO: This probably causes an obscure bug where confused players
         // going 'down' into the vestibule are twice as likely to fall, because
         // they have to pass a check here, and later in floor_transition
         // Right solution is probably to use the canonicalized direction everywhere
+
+        // If player falls down the stairs trying to leave the dungeon, we set
+        // the destination depth to 1 (i.e. D:1)
+        if (how == DNGN_EXIT_DUNGEON)
+            dest.depth = 1;
```

Comment:
I hope this is clear what I'm doing here and that this is OK. the issue is, when dest is returned as the default with branch = 0 and depth = -1, it's interpreted as invalid.  this is a problem because if we are inside this if block, it means the player slipped down the stairs and we want to prevent them from going_up.  This is OK for all up-stairs, EXCEPT leaving the dungeon. This is because the check to see if level transition should be blocked is special cased to only trigger when dest is invalid (i.e. the default) and the player is NOT trying to leave the dungeon.

Because of this, when slipping down the stairs when leaving the dungeon, I set the dest depth to 1 (which is actually unambiguously the actual destination in this situation). Then, this returned value is now valid and we can test for this condition to see if we need to block the transition.